### ISSUE TITLE
FINAL PROJECT

### ISSUE THREAD

**Issue by Prak2004** [Prak2004/Prak2004#1]
import pandas as pd 
import numpy as np 
from sklearn.model_selection import train_test_split 
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer 
from sklearn.naive_bayes import MultinomialNB 
from sklearn.metrics import accuracy_score, classification_report 
from nltk.corpus import stopwords 
from nltk.tokenize import word_tokenize 
from nltk.stem import PorterStemmer 
import nltk 
nltk.download('punkt') 
nltk.download('stopwords') 
# Load dataset 
url = "https://raw.githubusercontent.com/justmarkham/pycon-2016-tutorial/master/data/sms.tsv" 
data = pd.read_csv(url, sep='\t', header=None, names=['label', 'message']) 
# Map 'ham' to 0 and 'spam' to 1 
data['label'] = data['label'].map({'ham': 0, 'spam': 1}) 
# Initialize stopwords and stemmer 
stop_words = set(stopwords.words('english')) 
stemmer = PorterStemmer() 
def preprocess_text(text): 
# Convert to lowercase 
text = text.lower() 
# Tokenize 
tokens = word_tokenize(text) 
# Remove stopwords and apply stemming 
tokens = [stemmer.stem(word) for word in tokens if word.isalnum() and word not in stop_words] 
# Reconstruct the cleaned text 
return ' '.join(tokens) 
# Apply preprocessing 
data['cleaned_message'] = data['message'].apply(preprocess_text) 
# Initialize vectorizer 
vectorizer = TfidfVectorizer(max_features=5000) 
# Fit and transform the cleaned messages 
X = vectorizer.fit_transform(data['cleaned_message']).toarray() 
y = data['label'] 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42) 
# Train the model 
model = MultinomialNB() 
model.fit(X_train, y_train) 
# Predict on test data 
y_pred = model.predict(X_test) 
# Evaluate performance 
print("Accuracy:", accuracy_score(y_test, y_pred)) 
print("\nClassification Report:\n", classification_report(y_test, y_pred)) 
def predict_spam(message): 
cleaned_message = preprocess_text(message) 
vectorized_message = vectorizer.transform([cleaned_message]).toarray() 
prediction = model.predict(vectorized_message) 
return "Spam" if prediction == 1 else "Ham" 
# Test the model 
new_message = "Congratulations! You've won a $1000 Walmart gift card. Go to http://bit.ly/123456 to 
claim now." 
print(predict_spam(new_message))
